File: services/pipeline-build-service/src/models/job.rs
//! Job and JobState models. Mirrors `proto/pipeline/builds.proto`.
//!
//! `JobState` is the Foundry "Builds.md § Job states" vocabulary
//! literally — `WAITING`, `RUN_PENDING`, `RUNNING`, `ABORT_PENDING`,
//! `ABORTED`, `FAILED`, `COMPLETED`. The serde rename keeps the JSON
//! and the SQL CHECK constraint aligned with the proto enum names.

use std::fmt;
use std::str::FromStr;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum JobState {
    #[serde(rename = "WAITING")]
    Waiting,
    #[serde(rename = "RUN_PENDING")]
    RunPending,
    #[serde(rename = "RUNNING")]
    Running,
    #[serde(rename = "ABORT_PENDING")]
    AbortPending,
    #[serde(rename = "ABORTED")]
    Aborted,
    #[serde(rename = "FAILED")]
    Failed,
    #[serde(rename = "COMPLETED")]
    Completed,
}

impl JobState {
    pub const ALL: &'static [JobState] = &[
        JobState::Waiting,
        JobState::RunPending,
        JobState::Running,
        JobState::AbortPending,
        JobState::Aborted,
        JobState::Failed,
        JobState::Completed,
    ];

    pub fn as_str(&self) -> &'static str {
        match self {
            JobState::Waiting => "WAITING",
            JobState::RunPending => "RUN_PENDING",
            JobState::Running => "RUNNING",
            JobState::AbortPending => "ABORT_PENDING",
            JobState::Aborted => "ABORTED",
            JobState::Failed => "FAILED",
            JobState::Completed => "COMPLETED",
        }
    }

    pub fn is_terminal(&self) -> bool {
        matches!(
            self,
            JobState::Aborted | JobState::Failed | JobState::Completed
        )
    }
}

impl fmt::Display for JobState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl FromStr for JobState {
    type Err = UnknownJobState;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "WAITING" => Ok(JobState::Waiting),
            "RUN_PENDING" => Ok(JobState::RunPending),
            "RUNNING" => Ok(JobState::Running),
            "ABORT_PENDING" => Ok(JobState::AbortPending),
            "ABORTED" => Ok(JobState::Aborted),
            "FAILED" => Ok(JobState::Failed),
            "COMPLETED" => Ok(JobState::Completed),
            other => Err(UnknownJobState(other.to_string())),
        }
    }
}

#[derive(Debug, thiserror::Error)]
#[error("unknown job state: {0}")]
pub struct UnknownJobState(pub String);

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct Job {
    pub id: Uuid,
    pub rid: String,
    pub build_id: Uuid,
    pub job_spec_rid: String,
    pub state: String,
    pub output_transaction_rids: Vec<String>,
    pub state_changed_at: DateTime<Utc>,
    pub attempt: i32,
    pub stale_skipped: bool,
    pub failure_reason: Option<String>,
    pub output_content_hash: Option<String>,
    pub created_at: DateTime<Utc>,
}

impl Job {
    pub fn job_state(&self) -> Result<JobState, UnknownJobState> {
        JobState::from_str(&self.state)
    }
}

File: openfoundry-go/services/pipeline-build-service/internal/models/job.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// JobState mirrors Foundry "Builds.md § Job states" verbatim.
type JobState string

const (
	JobWaiting      JobState = "WAITING"
	JobRunPending   JobState = "RUN_PENDING"
	JobRunning      JobState = "RUNNING"
	JobAbortPending JobState = "ABORT_PENDING"
	JobAborted      JobState = "ABORTED"
	JobFailed       JobState = "FAILED"
	JobCompleted    JobState = "COMPLETED"
)

// AllJobStates lists every valid JobState — matches the Rust ALL slice.
var AllJobStates = []JobState{
	JobWaiting, JobRunPending, JobRunning, JobAbortPending,
	JobAborted, JobFailed, JobCompleted,
}

// IsTerminal mirrors the Rust `JobState::is_terminal`.
func (s JobState) IsTerminal() bool {
	return s == JobAborted || s == JobFailed || s == JobCompleted
}

// UnknownJobState is returned by ParseJobState on unknown input.
type UnknownJobState struct{ Value string }

func (e *UnknownJobState) Error() string { return "unknown job state: " + e.Value }

// ParseJobState converts a wire string to a typed JobState.
func ParseJobState(s string) (JobState, error) {
	for _, candidate := range AllJobStates {
		if string(candidate) == s {
			return candidate, nil
		}
	}
	return "", &UnknownJobState{Value: s}
}

// Job is the row shape for the `jobs` table.
type Job struct {
	ID                    uuid.UUID `json:"id"`
	RID                   string    `json:"rid"`
	BuildID               uuid.UUID `json:"build_id"`
	JobSpecRID            string    `json:"job_spec_rid"`
	State                 string    `json:"state"`
	OutputTransactionRIDs []string  `json:"output_transaction_rids"`
	StateChangedAt        time.Time `json:"state_changed_at"`
	Attempt               int32     `json:"attempt"`
	StaleSkipped          bool      `json:"stale_skipped"`
	FailureReason         *string   `json:"failure_reason,omitempty"`
	OutputContentHash     *string   `json:"output_content_hash,omitempty"`
	CreatedAt             time.Time `json:"created_at"`
}

// JobState projects the string column to a typed value.
func (j *Job) JobState() (JobState, error) { return ParseJobState(j.State) }
